### PR TITLE
Ruby layer refactor and new shadowenv layer

### PR DESCRIPTION
**Pull request by syl20bnr**
Related issue: https://github.com/syl20bnr/spacemacs/issues/14653

* new macro `spacemacs|eval-for-enabled-ruby-mode`
* use "dispatch" pattern with local vars hook for backend, test-runner, version-manager
* use spacebind
* diminish minor modes lighter
* new layer: shadowenv
* add shadowenv layer to ruby mode, python mode and compilation buffer

### CODE REVIEW

File: layers/+tools/shadowenv/README.org
Comment (lebensterben) [inline layers/+tools/shadowenv/README.org:5]: I suggest to write copyright notice for all image files from now on.

So add a footnote `[[file:img/shadowenv.png]] [fn:1]`
And `[fn:1] This logo image consists only of simple geometric shapes or text. It does not meet the threshold of originality needed for copyright protection, and is therefore in the public domain.`

File: layers/+tools/shadowenv/packages.el
Comment (lebensterben) [inline layers/+tools/shadowenv/packages.el:29]: optionally we can use
`:spacediminish " Ⓥ" " V"`

It takes 0 to 3 arguments, same as those of `spacemacs|diminish`.
When the first argument is omitted, it's inferred from the package name, which is `(intern (format "%s-mode" PKG-NAME))`.

File: layers/+tools/shadowenv/packages.el
Comment (lebensterben) [inline layers/+tools/shadowenv/packages.el:36]: A think this is a common case. `spacemacs|spacebind` should accept either a symbol or a list of symbols as for `MODE`.

File: layers/+lang/ruby/funcs.el
Comment (lebensterben) [inline layers/+lang/ruby/funcs.el:229]: the third argument to `add-hook` is `DEPTH`.
It defaults to 0, and any non-nil symbol implies 90.
So this hook is likely to be run after other hooks (-100 is first and 100 is last).

Is this really the behaviour we want?

File: layers/+lang/ruby/funcs.el
Comment (lebensterben) [inline layers/+lang/ruby/funcs.el:67]: `old` is dynamically scoped. It's better to make it a function argument.

File: layers/+lang/ruby/funcs.el
Comment (lebensterben) [inline layers/+lang/ruby/funcs.el:58]: This part can be simplified.
First we let-bound a list of modes based on whether `ruby-enable-enh-ruby-mode` is t.
```elisp
(let* ((rubymode (if ruby-enable-enh-ruby-mode
                     "enh-ruby"
                  "ruby")
       (pairs (cl-loop for old in '("mode" "hook" "local-vars-hook")
                       for new = (s-join "-" (list rubymode old))
                       collect (cons old new)))
 ..)
```
Then we can simply
```elisp
  (dolist (pair pairs)
     (cl-nsubstitute-if (cdr pair) 'spacemacs//ruby-substitute-predicate (car pair)))
```

Comment (lebensterben) [reply layers/+lang/ruby/funcs.el:58]: The real question is, how efficient is `cl-nsubstitute-if`.

File: layers/+tools/shadowenv/packages.el
Comment (JAremko) [reply layers/+tools/shadowenv/packages.el:36]: @lebensterben  Already planed :smile:   We have a bunch of layers doing this kind of templating.

Comment (syl20bnr) [reply layers/+tools/shadowenv/packages.el:36]: I did not think about static analysis. I guess we should drop the idea of static analysis because it will be like mission impossible to catch all the calls to `spacebind` correctly. For example here, even if I drop the macro to replace with a `dolist` (which would an inferior solution) the static analysis will have a hard time to detect each iteration of the `spacebind` call.
A better place for static analysis could be the `README.org` files once the key bindings have been rendered into them.

File: layers/+tools/shadowenv/packages.el
Comment (syl20bnr) [reply layers/+tools/shadowenv/packages.el:29]: Thanks for mentioning `:spacediminish`, I was not aware of it.

Done.

File: layers/+lang/ruby/funcs.el
Comment (lebensterben) [inline layers/+lang/ruby/funcs.el:177]: typo?

**Comment by syl20bnr** [comment]
I updated the conventions for toggles lighters in the mode line, see commit message https://github.com/syl20bnr/spacemacs/commit/c8ae548c9
These conventions were missing from the convention doc, this commit adds them.

File: layers/+tools/shadowenv/packages.el
Comment (syl20bnr) [reply layers/+tools/shadowenv/packages.el:36]: Makes me think that the macro here will prevent from correctly rendering in the `README.org`. 
You are right, passing an explicit list of modes is the way to do.

File: layers/+tools/shadowenv/README.org
Comment (Kazark) [inline layers/+tools/shadowenv/README.org:17]: typo: `activatefouhnd`

File: layers/+lang/ruby/funcs.el
Comment (syl20bnr) [reply layers/+lang/ruby/funcs.el:229]: I guess so. The addition of `DEPTH` is a rather new one and it is meant to be backward compatible. 
If there was `t` to append, then `DEPTH` should keep the same behavior.

File: layers/+lang/ruby/funcs.el
Comment (syl20bnr) [reply layers/+lang/ruby/funcs.el:67]: I thought about it but I could not do that, the predicate for `cl-substitute` is a function with one argument.
What could be done to improve upon this ? Sort of a curried function ? Or maybe declare variables explicitly ? Another macro ?

File: layers/+lang/ruby/funcs.el
Comment (syl20bnr) [reply layers/+lang/ruby/funcs.el:58]: Good question, I would assume it is the fastest as it is the destructive version `cl-substitute-if` but this is a wild guess 😅 
Going to update with your great implementation for now.

Comment (syl20bnr) [reply layers/+lang/ruby/funcs.el]: Done

File: layers/+lang/ruby/funcs.el
Comment (syl20bnr) [reply layers/+lang/ruby/funcs.el:166]: Fixed.

File: layers/+tools/shadowenv/README.org
Comment (syl20bnr) [reply layers/+tools/shadowenv/README.org]: Fixed.

File: layers/+tools/shadowenv/README.org
Comment (syl20bnr) [reply layers/+tools/shadowenv/README.org]: :+1: done.

File: layers/+tools/shadowenv/packages.el
Comment (syl20bnr) [reply layers/+tools/shadowenv/packages.el:36]: OK so we can keep it like it for now. When we support a list of modes we will migrate the spacebind calls out of the macro.

File: layers/+lang/ruby/funcs.el
Comment (lebensterben) [reply layers/+lang/ruby/funcs.el:56]: Solution one

well. let's make it to take pair, (OLD . BODY).
it's a function taking ONE argument.
`(cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate (cons old body)`

---

Make it take two arguments, the first is `old`
Solution two `(cl-nsubstitute-if new (apply-partially old) body)`.

**Comment by lebensterben** [comment]
Also it'd be nice if you can change
`(add-hook 'foo 'bar)` to `(add-hook 'foo #'bar)` where bar is a symbol of function.

This is not strictly enforced now but is easier for byte compiler to understand such symbol is a function.

**Review by smile13241324 — CHANGES_REQUESTED** [review]
Just reviewed your PR. I have found some smaller typos which should be changed before it can be merged.

**Comment by syl20bnr** [comment]
Do not merge this yet, I'm stil working on it.

File: layers/+spacemacs/spacemacs-defaults/keybindings.el
Comment (JAremko) [inline layers/+spacemacs/spacemacs-defaults/keybindings.el:741]: @syl20bnr  Extra set of `()`


@lebensterben  I think we should detect it and report as en error since it is easy to introduce when refactoring.

**Comment by smile13241324** [comment]
Given that this is now older than 3 years I would like to go ahead and close it for now.

**Pull request closed without merging**
